File: packages/sdk-python/syntra/tracing/span.py
"""Span implementation for Syntra SDK."""

from __future__ import annotations

import time
from abc import ABC, abstractmethod
from typing import Any

from syntra.tracing.context import (
    TRACE_FLAG_SAMPLED,
    SpanContext,
    generate_span_id,
    generate_trace_id,
)
from syntra.types import SpanEvent, SpanKind, SpanStatus, SpanStatusCode, TelemetrySpan


class Span(ABC):
    """Abstract base class for spans."""

    @property
    @abstractmethod
    def trace_id(self) -> str:
        """Get trace ID."""
        pass

    @property
    @abstractmethod
    def span_id(self) -> str:
        """Get span ID."""
        pass

    @property
    @abstractmethod
    def parent_span_id(self) -> str | None:
        """Get parent span ID."""
        pass

    @property
    @abstractmethod
    def name(self) -> str:
        """Get span name."""
        pass

    @abstractmethod
    def set_status(self, code: SpanStatusCode, message: str | None = None) -> None:
        """Set span status."""
        pass

    @abstractmethod
    def set_attribute(self, key: str, value: str | int | float | bool) -> None:
        """Set a single attribute."""
        pass

    @abstractmethod
    def set_attributes(self, attrs: dict[str, str | int | float | bool]) -> None:
        """Set multiple attributes."""
        pass

    @abstractmethod
    def add_event(
        self, name: str, attributes: dict[str, str | int | float | bool] | None = None
    ) -> None:
        """Add an event to the span."""
        pass

    @abstractmethod
    def end(self) -> None:
        """End the span."""
        pass

    @abstractmethod
    def is_recording(self) -> bool:
        """Check if span is still recording."""
        pass

    @abstractmethod
    def span_context(self) -> SpanContext:
        """Get span context for propagation."""
        pass

    def __enter__(self) -> Span:
        """Enter context manager."""
        return self

    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        """Exit context manager."""
        if exc_type is not None:
            self.set_status(
                SpanStatusCode.ERROR,
                str(exc_val) if exc_val else "Error",
            )
        self.end()


class SpanImpl(Span):
    """Span implementation."""

    def __init__(
        self,
        name: str,
        kind: SpanKind = SpanKind.INTERNAL,
        trace_id: str | None = None,
        parent_span_id: str | None = None,
        attributes: dict[str, str | int | float | bool] | None = None,
    ) -> None:
        self._name = name
        self._kind = kind
        self._trace_id = trace_id or generate_trace_id()
        self._span_id = generate_span_id()
        self._parent_span_id = parent_span_id
        self._start_time_ns = time.time_ns()
        self._end_time_ns: int | None = None
        self._status = SpanStatus()
        self._attributes: dict[str, str | int | float | bool] = attributes or {}
        self._events: list[SpanEvent] = []
        self._recording = True

    @property
    def trace_id(self) -> str:
        return self._trace_id

    @property
    def span_id(self) -> str:
        return self._span_id

    @property
    def parent_span_id(self) -> str | None:
        return self._parent_span_id

    @property
    def name(self) -> str:
        return self._name

    @property
    def kind(self) -> SpanKind:
        return self._kind

    @property
    def start_time_ns(self) -> int:
        return self._start_time_ns

    @property
    def duration_ns(self) -> int:
        if self._end_time_ns is None:
            return 0
        return self._end_time_ns - self._start_time_ns

    @property
    def status(self) -> SpanStatus:
        return self._status

    @property
    def attributes(self) -> dict[str, str | int | float | bool]:
        return dict(self._attributes)

    @property
    def events(self) -> list[SpanEvent]:
        return list(self._events)

    def set_status(self, code: SpanStatusCode, message: str | None = None) -> None:
        if not self._recording:
            return
        self._status = SpanStatus(code=code, message=message)

    def set_attribute(self, key: str, value: str | int | float | bool) -> None:
        if not self._recording:
            return
        self._attributes[key] = value

    def set_attributes(self, attrs: dict[str, str | int | float | bool]) -> None:
        if not self._recording:
            return
        self._attributes.update(attrs)

    def add_event(
        self, name: str, attributes: dict[str, str | int | float | bool] | None = None
    ) -> None:
        if not self._recording:
            return
        self._events.append(
            SpanEvent(
                name=name,
                timestamp_ns=time.time_ns(),
                attributes=attributes or {},
            )
        )

    def end(self) -> None:
        if not self._recording:
            return
        self._end_time_ns = time.time_ns()
        self._recording = False

    def is_recording(self) -> bool:
        return self._recording

    def span_context(self) -> SpanContext:
        return SpanContext(
            trace_id=self._trace_id,
            span_id=self._span_id,
            trace_flags=TRACE_FLAG_SAMPLED,
        )

    def to_telemetry_span(self, service_id: str, deployment_id: str) -> TelemetrySpan:
        """Convert to TelemetrySpan format for transport."""
        return TelemetrySpan(
            trace_id=self._trace_id,
            span_id=self._span_id,
            parent_span_id=self._parent_span_id,
            service_id=service_id,
            deployment_id=deployment_id,
            operation_name=self._name,
            span_kind=self._kind,
            start_time_ns=self._start_time_ns,
            duration_ns=self.duration_ns,
            status=self._status,
            attributes=self._attributes,
            events=self._events,
        )


class NoopSpan(Span):
    """No-op span for when sampling decides not to record."""

    def __init__(self, context: SpanContext | None = None) -> None:
        self._trace_id = context.trace_id if context else generate_trace_id()
        self._span_id = context.span_id if context else generate_span_id()
        self._name = "noop"

    @property
    def trace_id(self) -> str:
        return self._trace_id

    @property
    def span_id(self) -> str:
        return self._span_id

    @property
    def parent_span_id(self) -> str | None:
        return None

    @property
    def name(self) -> str:
        return self._name

    def set_status(self, code: SpanStatusCode, message: str | None = None) -> None:
        pass

    def set_attribute(self, key: str, value: str | int | float | bool) -> None:
        pass

    def set_attributes(self, attrs: dict[str, str | int | float | bool]) -> None:
        pass

    def add_event(
        self, name: str, attributes: dict[str, str | int | float | bool] | None = None
    ) -> None:
        pass

    def end(self) -> None:
        pass

    def is_recording(self) -> bool:
        return False

    def span_context(self) -> SpanContext:
        return SpanContext(
            trace_id=self._trace_id,
            span_id=self._span_id,
            trace_flags=0,
        )

File: packages/sdk-js/src/tracing/span.ts
import type {
  Span as ISpan,
  SpanContext,
  SpanKind,
  SpanStatus,
  SpanEvent,
  TelemetrySpan,
} from '../types';
import { generateSpanId, generateTraceId, TRACE_FLAG_SAMPLED } from './context';

/**
 * Span implementation
 */
export class SpanImpl implements ISpan {
  readonly traceId: string;
  readonly spanId: string;
  readonly parentSpanId?: string;
  readonly name: string;
  readonly startTime: number;
  readonly kind: SpanKind;

  private _status: SpanStatus = { code: 'unset' };
  private _attributes: Record<string, string | number | boolean> = {};
  private _events: SpanEvent[] = [];
  private _endTime: number | null = null;
  private _recording = true;

  constructor(options: {
    name: string;
    kind?: SpanKind;
    traceId?: string;
    parentSpanId?: string;
    attributes?: Record<string, string | number | boolean>;
  }) {
    this.name = options.name;
    this.kind = options.kind ?? 'internal';
    this.traceId = options.traceId ?? generateTraceId();
    this.spanId = generateSpanId();
    this.parentSpanId = options.parentSpanId;
    this.startTime = Date.now() * 1_000_000; // nanoseconds

    if (options.attributes) {
      this._attributes = { ...options.attributes };
    }
  }

  /**
   * Set span status
   */
  setStatus(status: SpanStatus): void {
    if (!this._recording) return;
    this._status = status;
  }

  /**
   * Set a single attribute
   */
  setAttribute(key: string, value: string | number | boolean): void {
    if (!this._recording) return;
    this._attributes[key] = value;
  }

  /**
   * Set multiple attributes
   */
  setAttributes(attrs: Record<string, string | number | boolean>): void {
    if (!this._recording) return;
    Object.assign(this._attributes, attrs);
  }

  /**
   * Add an event to the span
   */
  addEvent(
    name: string,
    attributes?: Record<string, string | number | boolean>
  ): void {
    if (!this._recording) return;

    this._events.push({
      name,
      timestamp_ns: Date.now() * 1_000_000,
      attributes: attributes ?? {},
    });
  }

  /**
   * End the span
   */
  end(): void {
    if (!this._recording) return;

    this._endTime = Date.now() * 1_000_000;
    this._recording = false;
  }

  /**
   * Check if span is still recording
   */
  isRecording(): boolean {
    return this._recording;
  }

  /**
   * Get span context for propagation
   */
  spanContext(): SpanContext {
    return {
      traceId: this.traceId,
      spanId: this.spanId,
      traceFlags: TRACE_FLAG_SAMPLED,
    };
  }

  /**
   * Get attributes
   */
  get attributes(): Record<string, string | number | boolean> {
    return { ...this._attributes };
  }

  /**
   * Get events
   */
  get events(): SpanEvent[] {
    return [...this._events];
  }

  /**
   * Get status
   */
  get status(): SpanStatus {
    return { ...this._status };
  }

  /**
   * Get duration in nanoseconds (0 if not ended)
   */
  get durationNs(): number {
    if (this._endTime === null) return 0;
    return this._endTime - this.startTime;
  }

  /**
   * Convert to TelemetrySpan format for transport
   */
  toTelemetrySpan(serviceId: string, deploymentId: string): TelemetrySpan {
    return {
      trace_id: this.traceId,
      span_id: this.spanId,
      parent_span_id: this.parentSpanId,
      service_id: serviceId,
      deployment_id: deploymentId,
      operation_name: this.name,
      span_kind: this.kind,
      start_time_ns: this.startTime,
      duration_ns: this.durationNs,
      status: this._status,
      attributes: this._attributes,
      events: this._events,
    };
  }
}

/**
 * No-op span for when sampling decides not to record
 */
export class NoopSpan implements ISpan {
  readonly traceId: string;
  readonly spanId: string;
  readonly parentSpanId?: string;
  readonly name: string;
  readonly startTime: number = 0;
  readonly kind: SpanKind = 'internal';

  constructor(context?: SpanContext) {
    this.traceId = context?.traceId ?? generateTraceId();
    this.spanId = context?.spanId ?? generateSpanId();
    this.name = 'noop';
  }

  setStatus(_status: SpanStatus): void {}
  setAttribute(_key: string, _value: string | number | boolean): void {}
  setAttributes(_attrs: Record<string, string | number | boolean>): void {}
  addEvent(_name: string, _attributes?: Record<string, string | number | boolean>): void {}
  end(): void {}
  isRecording(): boolean {
    return false;
  }
  spanContext(): SpanContext {
    return {
      traceId: this.traceId,
      spanId: this.spanId,
      traceFlags: 0,
    };
  }
}
